checker/src/main/java/org/checkerframework/checker/initialization/InitializationVisitor.java

```diff
@@ -358,48 +360,72 @@ protected void reportCommonAssignmentError(
         AnnotationMirror unknownInit = expected.getAnnotation(UnknownInitialization.class);
         AnnotationMirror underInit = expected.getAnnotation(UnderInitialization.class);
+
+        // If the actual receiver type (found) is not a subtype of expected,
+        // we still do not report an error if all necessary fields are initialized in the store
+        // before the method call.
+
+        // Find the frame for which the receiver must be initialized to discharge this error:
+        // * If the expected type is @UnknownInitialization(A) or @UnderInitialization(A), the frame
+        // is A.
+        // * If the expected type is @Initialized and the receiver type is final, the frame
+        // is the receiver type.
+        // * Otherwise, this error cannot be discharged and is reported by the super method.
         TypeMirror frame;
         if (unknownInit != null) {
             frame = atypeFactory.getTypeFrameFromAnnotation(unknownInit);
         } else if (underInit != null) {
             frame = atypeFactory.getTypeFrameFromAnnotation(underInit);
+        } else if (init != null
+                && ElementUtils.isFinal(TypesUtils.getTypeElement(expected.getUnderlyingType()))) {
+            frame = expected.getUnderlyingType();
         } else {
             super.reportMethodInvocabilityError(node, found, expected);
             return;
         }
 
-        TypeMirror receiverType = atypeFactory.getReceiverType(node).getUnderlyingType();
-        if (!atypeFactory.getProcessingEnv().getTypeUtils().isSubtype(frame, receiverType)) {
+        TypeMirror underlyingReceiverType = atypeFactory.getReceiverType(node).getUnderlyingType();
+        if (!atypeFactory
+                .getProcessingEnv()
+                .getTypeUtils()
+                .isSubtype(frame, underlyingReceiverType)) {
             super.reportMethodInvocabilityError(node, found, expected);
             return;
         }
 
+        GenericAnnotatedTypeFactory<?, ?, ?, ?> targetFactory =
+                checker.getTypeFactoryOfSubcheckerOrNull(
+                        ((InitializationChecker) checker).getTargetCheckerClass());
         List<VariableTree> uninitializedFields =
                 atypeFactory.getUninitializedFields(
                         atypeFactory.getStoreBefore(node),
+                        targetFactory.getStoreBefore(node),
                         getCurrentPath(),
                         false,
                         Collections.emptyList());
         uninitializedFields.removeAll(initializedFields);
 
-        filterInitializedFields(node, uninitializedFields, true);
-
         if (!uninitializedFields.isEmpty()) {
-            StringJoiner fieldsString = new StringJoiner(", ");
-            for (VariableTree f : uninitializedFields) {
-                fieldsString.add(f.getName());
-            }
+            // TODO: improve the error message by showing the uninitialized fields
```

Comment:
@Ao-senXiong Can you please open an issue for this TODO and see how to improve this error message?

Reply:
> @Ao-senXiong Can you please open an issue for this TODO and see how to improve this error message?

Ok, I will open up an issue